### ISSUE TITLE
Setting other hyperparameters like C and Weight

### ISSUE THREAD

**Issue by harshithnarra** [svm-abstract-verifier/saver#6]
My SVM model has other hyperparameters like C and weights of the classes, how can I set these hyperparameters.
Is there a way?

_Originally posted by @harshithnarra in https://github.com/abstract-machine-learning/saver/issues/5#issuecomment-823380176_

**Comment by marco-zanella** [comment]
Our SVM trainer at [trainers/svm.py](https://github.com/abstract-machine-learning/data-collection/blob/master/trainers/svm.py) is quite limited in that sense, it acts more like a façade.
You can however train your SVM in Python as you normally would, then pass the resulting model to ClassifierMapper.create ([trainers/classifier_mapper.py](https://github.com/abstract-machine-learning/data-collection/blob/master/trainers/classifier_mapper.py)). It will export your model as long as it is an instance of [sklearn.svm.SVC](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html), since hyperparameters like C are relevant during the training phase only and need not to be exported.

**Comment by harshithnarra** [comment]
My dataset is unbalanced( there are lot more cases labelled as 0 when compared to cases labelled as 1).
I am using the following model for training:
weights = {1:6.5, 0:1.0}
model=SVC(C=50.34933990539188,kernel='poly',class_weight=weights)

When I train this model and test it against test set containing 19 cases(16 cases labelled as 0 and 3 cases labelled as 1) I get the following confusion matrix:
[[12  4]
 [ 0  3]]

But when I run SAVer with the same test set all cases are predicted to be 0 which is
[[16 0]
 [3   0]]

Is there something that I can do? Could you please guide me.

**Comment by marco-zanella** [comment]
It looks like the problem is not related to class weights, but to the _gamma_ parameter used by sklearn. The [sklearn documentation](https://scikit-learn.org/stable/modules/generated/sklearn.svm.SVC.html) states that "gamma is a kernel coefficient for [...] poly", which shouldn't be, according to the standard definition of [Polynomial Kernel](https://en.wikipedia.org/wiki/Polynomial_kernel). Training polynomial SVMs using different values for _gamma_ yields different results, hence my understanding is that sklearn somehow uses this parameter.

I found a related question on [Stack Exchange](https://stats.stackexchange.com/questions/168014/where-can-i-read-about-gamma-coefficient-in-svm-in-scikit-learn) where author asks about the meaning of gamma. He gets an answer for _RBF_, but no one seems to know why or how it is used for _poly_.

Unfortunately all of this appears to be undocumented, so we don't know how to reproduce sklearn's behavior.

**Comment by harshithnarra** [comment]
I tried the same with linear and RBF models. The prediction for the test set from sklearn(python) is exactly the opposite of what SAVer predicts. Both the predictions are complements of each other.
For example, if the prediction of sklearn is:
 [0. 0. 0. 0. 1. 0. 0. 1. 0. 1. 1. 0. 1. 1. 1. 1. 1. 1. 0.]
SAVer predicts:
 [1. 1. 1. 1. 0. 1. 1. 0. 1. 0. 0. 1. 0. 0. 0. 0. 0. 0. 1.]

Can you please help me figure out what the minor fix could be? As in every case, the predictions are exact complements of each other. I ensured that the labelling is not mixed up in either the data file or the model file.

**Comment by marco-zanella** [comment]
This is likely to be an issue in the classifier_mapper.py swapping labels before export. A quick fix would be to manually edit the SVM .dat file and swapping the two labels back, i.e. from this:
ovo 7 2
rbf 1
0 8
1 3
...

to:
ovo 7 2
rbf 1
1 8
0 3

I will fix the issue in the classifier mapper meanwhile.

**Comment by marco-zanella** [comment]
Fixed in [https://github.com/abstract-machine-learning/data-collection/commit/8e4a9ab0c720e19d4641922f3dd09c80286d6155](https://github.com/abstract-machine-learning/data-collection/commit/8e4a9ab0c720e19d4641922f3dd09c80286d6155)
